[SPARK-29971][CORE] Fix buffer leaks in `TransportFrameDecoder/TransportCipher`

normanmaurer opened the pull request:
### What changes were proposed in this pull request?

- Correctly release `ByteBuf` in `TransportCipher` in all cases
- Move closing / releasing logic to `handlerRemoved(...)` so we are guaranteed that is always called.
- Correctly release `frameBuf` it is not null when the handler is removed (and so also when the channel becomes inactive)

### Why are the changes needed?

We need to carefully manage the ownership / lifecycle of `ByteBuf` instances so we don't leak any of these. We did not correctly do this in all cases:
 - when end up in invalid cipher state.
 - when partial data was received and the channel is closed before the full frame is decoded

Fixes https://github.com/netty/netty/issues/9784.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Pass the newly added UTs.

SparkQA commented:
**[Test build #114150 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114150/testReport)** for PR 26609 at commit [`a8420ee`](https://github.com/apache/spark/commit/a8420ee7b2e6c2568773407667aafd03d0681fad).

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #114151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114151/testReport)** for PR 26609 at commit [`c766150`](https://github.com/apache/spark/commit/c766150613d95dde080680134b6c2711662e98c0).

SparkQA commented:
**[Test build #114150 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114150/testReport)** for PR 26609 at commit [`a8420ee`](https://github.com/apache/spark/commit/a8420ee7b2e6c2568773407667aafd03d0681fad).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114150/
Test FAILed.

SparkQA commented:
**[Test build #114155 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114155/testReport)** for PR 26609 at commit [`bf1299b`](https://github.com/apache/spark/commit/bf1299be9b8fb899e624c58640d4421aab3ff4e0).

normanmaurer committed:
[SPARK-29971] Fix multiple possible buffer leaks in `TransportFrameDecoder` and `TransportCipher`

Motivation:

We need to carefully manage the ownership / lifecycle of `ByteBuf` instances so we don't leak any of these. We did not correctly do this in all cases:
 - when end up in invalid cipher state.
 - when partial data was received and the channel is closed before the full frame is decoded

Modifications:

- Release message in finally block to ensure it is also released in case of cipher is not valid
- Move closing / releasing logic to `handlerRemoved(...)` so we are guarenteed that is always called.
- Add tests for `TransportCipher`
- Correctly release `frameBuf` it is not null when the handler is removed (and so also when the channel becomes inactive)

Result:

Fixes https://github.com/netty/netty/issues/9784.

SparkQA commented:
**[Test build #114159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114159/testReport)** for PR 26609 at commit [`d35f381`](https://github.com/apache/spark/commit/d35f381f6816df74443e3f320db2132d442d36c8).

SparkQA commented:
**[Test build #114151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114151/testReport)** for PR 26609 at commit [`c766150`](https://github.com/apache/spark/commit/c766150613d95dde080680134b6c2711662e98c0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114151/
Test FAILed.

SparkQA commented:
**[Test build #114155 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114155/testReport)** for PR 26609 at commit [`bf1299b`](https://github.com/apache/spark/commit/bf1299be9b8fb899e624c58640d4421aab3ff4e0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114155/
Test PASSed.

SparkQA commented:
**[Test build #114159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114159/testReport)** for PR 26609 at commit [`d35f381`](https://github.com/apache/spark/commit/d35f381f6816df74443e3f320db2132d442d36c8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114159/
Test PASSed.

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -19,23 +19,64 @@
+    if (currentData == null) {
+      data = buf;
+    } else {
+      int currentReadable = currentData.readableBytes();
```
It seems you're making this change so that it's safe to call `feedData` when there is still data to process.

That's unnecessary; this class is not general purpose, and is only used in one place, where the surrounding code makes sure all this data is consumed after `feedData` is called. If you're worried, a check for the buffer being empty + throwing an exception should be enough.

(I'd kinda like to rewrite this stuff at some point to avoid so much copying and add new features, but that requires time...)

normanmaurer replied:
@vanzin the class is public and as it is today very error-prone. What happens if someone start using it from another place ? So either we should throw or change the class to be safe to use (which I did here). If you prefer throwing that's fine as well.

vanzin replied:
It's not a public API for Spark. Spark has a lot of public classes that are not considered part of its public API. People shouldn't use those except at their own risk.

(Only the classes in Spark's public API docs are considered really public, and this whole module is not there.)

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -45,18 +86,24 @@ public int read(ByteBuffer dst) throws IOException {
   public void close() throws IOException {
+    closed = true;
+    if (data != null) {
+      data.release();
```
This feels like the wrong place to do it. Shouldn't `TransportCipher.channelRead` do that instead after it decrypts the buffer? (Then you also don't need L46.)

normanmaurer replied:
@vanzin again this class is public and so I think it is not guaranteed that read(...) will always consume the whole buffer.  The `ByteArrayReadableChannel` is wrapped in a `CryptoInputStream` and who knows if it always is guaranteed that it will consume all data (for example there is at least one case where you catch an `InternalError` which may cause that the previous feed data was not released (as it was not fully consumed).

So with all this in mind I think it is the safest to at least ensure we release everything once `close()` is called.

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/util/TransportFrameDecoder.java:
```diff
@@ -184,8 +184,12 @@ private ByteBuf decodeNext() {
     }
 
-    // Reset buf and size for next frame.
+    return comsumeCurrentFrameBuf();
```
consume

normanmaurer replied:
ups typo... I blame my fat fingers :) Let me fix this.

vanzin reviewed (commented):
`ByteArrayReadableChannel` is kind of a weird class, I'd rather make the (only) caller do the right thing instead.

dongjoon-hyun commented:
Thank you for making a PR, @normanmaurer !

normanmaurer commented:
@vanzin also related to your comment of `ByteArrayReadableChannel`.... imho you transfer ownership of the `ByteBuf` to `ByteArrayReadableChannel` so it should be its responsibility to ensure we release it.

vanzin commented:
> you transfer ownership of the ByteBuf to ByteArrayReadableChannel

As I mentioned that's a weird class. It's meant to be fed the buffer, and the caller is expected to consume all the data right away, as `TransportCipher` does. So technically `TransportCipher` never loses ownership of that buffer, and this class is just an optimization to avoid instantiating a new one every time data arrives.

If you're really so worried about all this you could make this class a private static class inside `TransportCipher` which is the only place where it's used.

normanmaurer commented:
@vanzin like mentioned above I am not sure it is safe to assume it always consume all data as an `InternalError` may be thrown (you even guard against it). So I would suggest to be better safe then sorry.  If you are super concerned about the extra logic in `feedData(...)` I would also be happy to let it throw if we already have a reference to a `ByteBuf`. Just let me know if you would prefer this and I will change the code. That said I am not sure why it is such a big deal to add the code to `feedData` as I did as I don't think it is overly complex. But I am happy to adapt if needed :)

normanmaurer commented:
@dongjoon-hyun @vanzin I just added another commit to add a test case related to your handling of `InternalError` which would have leaked two buffers before these changes in here.

SparkQA commented:
**[Test build #114238 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114238/testReport)** for PR 26609 at commit [`b06ec10`](https://github.com/apache/spark/commit/b06ec107b07483b9a4fb6e803fecd9794a519398).

SparkQA commented:
**[Test build #114238 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114238/testReport)** for PR 26609 at commit [`b06ec10`](https://github.com/apache/spark/commit/b06ec107b07483b9a4fb6e803fecd9794a519398).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114238/
Test FAILed.

normanmaurer commented:
@vanzin @dongjoon-hyun can you let me know how I can ignore the linter for the two "InternalError" instances that are thrown by the testcode ?

normanmaurer commented:
@vanzin I am not talking about the changes in `feedData(...)` here (which I think we can also just let it throw an exception as stated above). I am saying that if we not release the buffer in `close()` we may leak as there is no guarantee that we consume the whole buffer and so release it if an `InternalError` was thrown.

vanzin commented:
Still not following you.

`feedData` is called once in `channelRead`, and the contract in the code is that all that buffer's data must be read in that same call to `channelRead`. So you can release the buffer in `channelRead` just fine as far as I can see. `close` doesn't need to do anything.

normanmaurer commented:
@vanzin yes we could also just remove all the `release` magic in `ByteArrayReadableChannel.read(...)` and `close(...)` and handle all in `channelRead(...)` that would be fine with me as well. When I started the patch my assumption was that `ByteArrayReadableChannel` may be used by users and so we should not do this because of this. If this not a problem then sure lets do it. Please let me know.

dongjoon-hyun commented on common/network-common/src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:
```diff
@@ -0,0 +1,110 @@
+          @Override
+          public int read(ByteBuffer dst) {
+            // Simulate throwing InternalError
+            throw new InternalError();
```
@normanmaurer . `dev/lint-java` complains at this line and line 76.
```
[ERROR] src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:[70] (regexp) RegexpSinglelineJava: Avoid throwing error in application code.
[ERROR] src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:[76] (regexp) RegexpSinglelineJava: Avoid throwing error in application code.
```

normanmaurer replied:
I know but I want to throw this as the test actually test exactly for this scenario ;)

dongjoon-hyun replied:
Can we use `*Exception` instead in this test code?

vanzin replied:
The commons-crypto library throws `InternalError` so to test that behavior you have to use that error.

vanzin replied:
Found somewhere that you can use `// CHECKSTYLE:OFF` and `// CHECKSTYLE:ON` comments to disable checkstyle for some code.

dongjoon-hyun replied:
Got it. +1 for @vanzin 's advice.

normanmaurer replied:
Just pushed this change... lets see if it works :)

vanzin commented:
> remove all the release magic in ByteArrayReadableChannel.read(...) and close(...) and handle all in channelRead(...)

That sounds good to me.

normanmaurer commented:
@vanzin done..

SparkQA commented:
**[Test build #114251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114251/testReport)** for PR 26609 at commit [`34440e8`](https://github.com/apache/spark/commit/34440e8b17d8aaf4aa51a8f9dea7fe52e455ae5a).

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -19,44 +19,54 @@
+    if (closed) {
+      throw new ClosedChannelException();
+    }
+    if (data != null) {
```
Where are you setting `data` to null?

(In the spirit of keeping this class as small and dumb as possible, it might be better to check that `buffer.readableBytes()` is 0 before returning from `channelRead`. Then you may even be able to get rid of the `readableBytes()` method in this class.)

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -19,44 +19,54 @@
 
   @Override
   public int read(ByteBuffer dst) throws IOException {
+    if (closed) {
```
This is ok by itself, but I don't see any code calling `close()` so it's basically dead code...

normanmaurer replied:
I can remove it if you like but that is basically the contract of a `ReadableByteChannel`.

vanzin replied:
I think is fine. I took another look and `close()` is probably being called as part of closing the `CryptoInputStream`.

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -19,44 +19,54 @@
+      throw new ClosedChannelException();
+    }
+    if (data == null) {
+      return 0;
```
I'd throw an `IllegalStateException` instead.

SparkQA commented:
**[Test build #114251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114251/testReport)** for PR 26609 at commit [`34440e8`](https://github.com/apache/spark/commit/34440e8b17d8aaf4aa51a8f9dea7fe52e455ae5a).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114251/
Test FAILed.

vanzin commented on common/network-common/src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:
```diff
@@ -0,0 +1,110 @@
+                null, new IvParameterSpec(new byte[0])) {
+          @Override
+          public int read(ByteBuffer dst) {
+            // Simulate throwing InternalError
```
nit: these comments are redundant. If you want to comment that this mimics the `commons-crypto` library then it would be better, otherwise just remove them.

vanzin commented on common/network-common/src/test/java/org/apache/spark/network/util/ByteArrayReadableChannelTest.java:
```diff
@@ -0,0 +1,89 @@
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class ByteArrayReadableChannelTest {
```
I'm not seeing much value in these tests (especially given my latest comments). Especially the ref counting checks which now are not done by the `ByteArrayReadableChannel` at all...

SparkQA commented:
**[Test build #114252 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114252/testReport)** for PR 26609 at commit [`b4b1b76`](https://github.com/apache/spark/commit/b4b1b76c2928c1424a57a0eb997866aa43e7870a).

vanzin commented on common/network-common/src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:
```diff
@@ -0,0 +1,110 @@
+    String algorithm = "TestAlgorithm";
+    TransportConf conf = new TransportConf("Test", MapConfigProvider.EMPTY);
+    TransportCipher cipher = new TransportCipher(conf.cryptoConf(), conf.cipherTransformation(),
+            new SecretKeySpec(new byte[256], algorithm), new byte[0], new byte[0]) {
```
nit: indented too far (use 4 spaces here)

vanzin commented on common/network-common/src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:
```diff
@@ -0,0 +1,110 @@
+
+      @Override
+      CryptoInputStream createInputStream(ReadableByteChannel ch) throws IOException {
+        CryptoCipher cipher = mock(CryptoCipher.class);
```
Can't you mock `CryptoInputStream` directly?

`when(mockInputStream.read(any()).thenThrow(new InternalError())`?

normanmaurer commented:
@vanzin can you have a look again ?

dongjoon-hyun commented on common/network-common/src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:
```diff
@@ -0,0 +1,114 @@
+            // Simulate throwing InternalError
+            // CHECKSTYLE:OFF
+            throw new InternalError();
+            // CHECKSTYLE:ON
```
Oops. Sorry, we changed the accepted style. Please change like the following. I double-checked the following is working.
```java
            // checkstyle.off: RegexpSinglelineJava
             throw new InternalError();
            // checkstyle.on: RegexpSinglelineJava
```

normanmaurer replied:
I think I not need this anymore as I just use mocking now. Thanks anyway :)

dongjoon-hyun commented:
Oh, you changed to `Mock`. Then, please ignore my outdated comments~

SparkQA commented:
**[Test build #114253 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114253/testReport)** for PR 26609 at commit [`02edc2c`](https://github.com/apache/spark/commit/02edc2c11e004f7e624b297ba44f5fc603ffd6d0).

SparkQA commented:
**[Test build #114253 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114253/testReport)** for PR 26609 at commit [`02edc2c`](https://github.com/apache/spark/commit/02edc2c11e004f7e624b297ba44f5fc603ffd6d0).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114253/
Test FAILed.

vanzin commented on common/network-common/src/test/java/org/apache/spark/network/crypto/TransportCipherTest.java:
```diff
@@ -0,0 +1,99 @@
+
+      @Override
+      CryptoInputStream createInputStream(ReadableByteChannel ch) throws IOException {
+        CryptoCipher cipher = mock(CryptoCipher.class);
```
This variable is not needed anymore.

vanzin commented on common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java:
```diff
@@ -166,34 +167,46 @@ boolean isCipherValid() {
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+      // We do the closing of the stream / channel in handlerRemoved(...) as
+      // this method will be called
+      // in all cases:
```
fits in previous line

vanzin reviewed (commented):
Other than the style issues the linter picked up, looks ok.

normanmaurer commented:
@vanzin I hope I fixed all now.

SparkQA commented:
**[Test build #114254 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114254/testReport)** for PR 26609 at commit [`c745649`](https://github.com/apache/spark/commit/c745649c1074ca89ce66d6e8be49657e94fb50f5).

SparkQA commented:
**[Test build #114252 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114252/testReport)** for PR 26609 at commit [`b4b1b76`](https://github.com/apache/spark/commit/b4b1b76c2928c1424a57a0eb997866aa43e7870a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114252/
Test FAILed.

SparkQA commented:
**[Test build #114254 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114254/testReport)** for PR 26609 at commit [`c745649`](https://github.com/apache/spark/commit/c745649c1074ca89ce66d6e8be49657e94fb50f5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114254/
Test PASSed.

vanzin commented:
Merging to master / 2.4. I'll fix the import order in the new test during merge.

vanzin commented:
Actually, while merging I noticed something else... could you name the new test class `TransportCipherSuite` to match all the other test suites?

(While there, the import order convention is to have java/javax imports first.)

dongjoon-hyun commented:
So, this is still under review status for those comments, right? @vanzin

vanzin commented:
Yes, as you can see by the fact that it's still open...

SparkQA commented:
**[Test build #114280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114280/testReport)** for PR 26609 at commit [`d8d2bbe`](https://github.com/apache/spark/commit/d8d2bbedd5a09209ec305d528d0e7bed63d4621a).

normanmaurer commented:
@vanzin @dongjoon-hyun addressed all + squashed and adjusted commit message to reflect the final patch.

SparkQA commented:
**[Test build #114281 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114281/testReport)** for PR 26609 at commit [`fb79d0a`](https://github.com/apache/spark/commit/fb79d0ac15c6f572ec3d26fde4e73727cff95a9d).

SparkQA commented:
**[Test build #114281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114281/testReport)** for PR 26609 at commit [`fb79d0a`](https://github.com/apache/spark/commit/fb79d0ac15c6f572ec3d26fde4e73727cff95a9d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #114280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114280/testReport)** for PR 26609 at commit [`d8d2bbe`](https://github.com/apache/spark/commit/d8d2bbedd5a09209ec305d528d0e7bed63d4621a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class TransportCipherSuite `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114281/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114280/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #114306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114306/testReport)** for PR 26609 at commit [`fb79d0a`](https://github.com/apache/spark/commit/fb79d0ac15c6f572ec3d26fde4e73727cff95a9d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19146/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM.

SparkQA commented:
**[Test build #114306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114306/testReport)** for PR 26609 at commit [`fb79d0a`](https://github.com/apache/spark/commit/fb79d0ac15c6f572ec3d26fde4e73727cff95a9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114306/
Test PASSed.

srowen commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -19,44 +19,44 @@
 
 public class ByteArrayReadableChannel implements ReadableByteChannel {
   private ByteBuf data;
+  private boolean closed;
```
Not sure, does this need to be volatile? I'm not sure if there are thread safety concerns here.

vanzin replied:
streams / channels are not thread-safe.

srowen commented on common/network-common/src/main/java/org/apache/spark/network/util/ByteArrayReadableChannel.java:
```diff
@@ -19,44 +19,44 @@
     }
 
-    if (data.readableBytes() == 0) {
-      data.release();
```
Just checking this is no longer needed because of the new cleanup?

vanzin commented:
Norman, it's not clear whether you wanted your last commit message to be the PR description? Github does not do that automatically.

normanmaurer commented:
I think the PR description is fine 

> Am 22.11.2019 um 20:16 schrieb Marcelo Vanzin <notifications@github.com>:
> 
> ﻿
> Norman, it's not clear whether you wanted your last commit message to be the PR description? Github does not do that automatically.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

vanzin commented:
Merging to master / 2.4.

vanzin commented:
2.4 does not have SPARK-26674 so the cherry-pick was not clean (and it's not a super trivial fix). So need to either cherry-pick that change too, or post a fixed PR for 2.4, if you think this is important in that branch.

normanmaurer commented:
@vanzin let me have a look ... imho this fix is important as it could result in memory leaks

dongjoon-hyun commented:
Thank you, @vanzin and @normanmaurer .
Yes. Please create a PR against `branch-2.4` for `branch-2.4` memory leak, @normanmaurer .

normanmaurer commented:
@dongjoon-hyun @vanzin  done https://github.com/apache/spark/pull/26660
